module/zfs/zfs_vnops.c

```diff
@@ -3097,6 +3175,13 @@ zfs_setattr(struct inode *ip, vattr_t *vap, int flags, cred_t *cr)
 		SA_ADD_BULK_ATTR(xattr_bulk, xattr_count,
 		    SA_ZPL_FLAGS(zfsvfs), NULL, &attrzp->z_pflags,
 		    sizeof (attrzp->z_pflags));
+		if (projid != ZFS_INVALID_PROJID &&
+		    projid != attrzp->z_projid) {
+			attrzp->z_projid = projid;
+			SA_ADD_BULK_ATTR(xattr_bulk, xattr_count,
+			    SA_ZPL_PROJID(zfsvfs), NULL, &attrzp->z_projid,
+			    sizeof (attrzp->z_projid));
```

Comment:
Good question!
I need more handle here: if the object has no PROJID flag, then it should be existing file that is not upgraded yet, so we need to handle it as zfs_sa_upgrade() does. Does it work?